vertx-web-client/src/main/java/io/vertx/ext/web/client/CachingWebClient.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2021 Red Hat, Inc.
+ *
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Public License v1.0
+ *  and Apache License v2.0 which accompanies this distribution.
+ *
+ *  The Eclipse Public License is available at
+ *  http://www.eclipse.org/legal/epl-v10.html
+ *
+ *  The Apache License v2.0 is available at
+ *  http://www.opensource.org/licenses/apache2.0.php
+ *
+ *  You may elect to redistribute this code under either of these licenses.
+ */
+package io.vertx.ext.web.client;
+
+import io.vertx.codegen.annotations.GenIgnore;
+import io.vertx.codegen.annotations.VertxGen;
+import io.vertx.core.Vertx;
+import io.vertx.core.http.HttpClient;
+import io.vertx.ext.web.client.impl.CachingWebClientImpl;
+import io.vertx.ext.web.client.impl.cache.SharedDataCacheStore;
+import io.vertx.ext.web.client.spi.CacheStore;
+
+/**
+ * An asynchronous cache aware HTTP / HTTP/2 client called {@code CachingWebClient}.
+ * <p>
+ * This client wraps a {@link WebClient} and makes it cache aware by adding features to it:
+ * <ul>
+ *   <li>Cache-Control header parsing</li>
+ *   <li>Response serialization and deserialization for the {@link CacheStore}</li>
+ *   <li>Freshness checking</li>
+ * </ul>
+ * <p>
+ * The client honors the cache headers:
+ * <ul>
+ *  <li>Cache-Control with the following properties understood:
+ *    <ul>
+ *      <li>public</li>
+ *      <li>private</li>
+ *      <li>no-cache</li>
+ *      <li>no-store</li>
+ *      <li>max-age</li>
+ *      <li>s-maxage</li>
+ *    </ul>
+ *  </li>
+ *  <li>Expires</li>
+ *  <li>ETag</li>
+ *  <li>Vary</li>
+ * </ul>
+ * <p/>
+ *
+ * @author <a href="mailto:craigday3@gmail.com">Craig Day</a>
+ */
+@VertxGen
+public interface CachingWebClient {
+
+  /**
+   * Create a cache aware web client using an
+   * {@link io.vertx.core.shareddata.AsyncMap} as the {@link CacheStore}.
+   *
+   * @param vertx the vertx instance
+   * @return the creatd web client
+   */
+  static WebClient create(Vertx vertx) {
```

Comment:
We should agree on a factory default. Session aware and OAuth2 aware clients always wrap a preconfigured base web client in a method named `create(...)` and we don't have a `wrap(...)` variant. 

If we decide that we should simplify it for users to have a `create()` to create a base client and wrap, then we should also apply the same style to the other wrappers we already have.

IMHO, we are favouring composition over inheritance, so my personal choice (which is totally open for discussion) would be to remove the current `create()` factories and rename the `wrap(...)` factories to `create(...)` which would make it "consistent" with the other clients

Reply:
/cc @vietj for the comment above

Reply:
I agree with you @pmlopes

Reply:
@pmlopes and @vietj , just to be clear here do we want `CachingWebClient` to be just like `WebClientSession`? Meaning it should `extends WebClient`, or just renaming the static interface methods to match name/signature? Related, should the caching client implementation then `extends WebClientBase`, or can it just keep the two methods it has where it just adds the interceptor?

Reply:
@craig-day, no. Your PR is ok, what I recommend is a doing 2 small changes:

1. remove the all `create` methods
2. rename the remaining `wrap` to `create`

For example it should end up looking like this:

```java
package io.vertx.ext.web.client;

...

@VertxGen
public interface CachingWebClient {

  /**
   * Create a {@link HttpClient} with a web client and default options.
   *
   * @param httpClient the http client to wrap
   * @param cacheStore the cache adapter
   * @return the created web client
   */
  @GenIgnore(GenIgnore.PERMITTED_TYPE)
  static WebClient create(HttpClient httpClient, CacheStore cacheStore) {
    CachingWebClientOptions options = new CachingWebClientOptions();
    return wrap(httpClient, cacheStore, options);
  }

  /**
   * Create a {@link HttpClient} with a web client and custom options.
   *
   * @param httpClient the http client to wrap
   * @param cacheStore the cache adapter
   * @param options    the caching web client options
   * @return the created web client
   */
  @GenIgnore(GenIgnore.PERMITTED_TYPE)
  static WebClient create(HttpClient httpClient, CacheStore cacheStore, CachingWebClientOptions options) {
    return CachingWebClientImpl.wrap(httpClient, cacheStore, options);
  }
}
```

As this is the same API style already used by the WebClientOAuth2: https://github.com/vert-x3/vertx-web/blob/master/vertx-web-client/src/main/java/io/vertx/ext/web/client/WebClientOAuth2.java

Finally, we may want to agree on the interface name, so it also means we should rename the new `WebClientOAuth2` interface to `OAuth2WebClient` to keep the naming consistent:

* WebClient (base)
* CachingWebClient (this)
* OAuth2WebClient (the other wrapper added only in master atm)

Reply:
Thanks for the details @pmlopes. I cleaned up the in interface to 4 `create` methods. I think it's nice to have a method where you don't have to give a store and a default in-memory store is used. If we don't want this, I can slim the interface down to just the two methods you suggested. /cc @vietj